File: structural/C++/Bridge.cpp
#include <iostream>
#include <string>

class Fuel
{
public:
	virtual std::string type() = 0;
};

class Diesel : public Fuel
{
public:
	std::string type()
	{
		return "Diesel";
	}
};

class Gasoline : public Fuel
{
public:
	std::string type()
	{
		return "Gasoline";
	}
};

class Car
{
public:
	virtual void info() = 0;
};

class Jeep : public Car
{
public:
	Jeep(Fuel* fuel) : m_Fuel(fuel)
	{
	}

	~Jeep()
	{
		delete m_Fuel;
	}

	void info()
	{
		std::cout << "\nModel: Jeep" << std::endl;
		std::cout << "Fuel: " + m_Fuel->type() << std::endl;
	}

private:
	Fuel* m_Fuel;
};

class Subaru : public Car
{
public:
	Subaru(Fuel* fuel) : m_Fuel(fuel)
	{
	}

	~Subaru()
	{
		delete m_Fuel;
	}

	void info()
	{
		std::cout << "\nModel: Subaru" << std::endl;
		std::cout << "Fuel: " + m_Fuel->type() << std::endl;
	}

private:
	Fuel* m_Fuel;
};

int main()
{
	std::cout << ":: Bridge ::" << std::endl;

	Fuel* diesel = new Diesel();
	Fuel* gasoline = new Gasoline();

	Car* jeep = new Jeep(diesel);
	Car* subaru = new Subaru(gasoline);

	jeep->info();
	subaru->info();

	system("PAUSE");

	delete jeep;
	delete subaru;

	return 0;
}

File: structural/javascript/Bridge.js
// Interface
class Fuel {

  constructor() {
    if (this.constructor === Fuel) {
      throw new TypeError('Cannot construct Abstract instances directly');
    }

    if (this.type === Fuel.prototype.type) {
      throw new TypeError('Must override method "type()"');
    }
  }

  type() {
    throw new TypeError('Do not call abstract method "type()" from child.');
  }

}

class Diesel extends Fuel {

  // Implements abstract parent method
  type() {
    return "Diesel";
  }

}

class Gasoline extends Fuel {

  // Implements abstract parent method
  type() {
    return "Gasoline";
  }

}

// Interface
class Car {

  constructor() {
    if (this.constructor === Car) {
      throw new TypeError('Cannot construct Abstract instances directly');
    }

    if (this.showInfo === Car.prototype.showInfo) {
      throw new TypeError('Must override method "showInfo()"');
    }
  }

  showInfo() {
    throw new TypeError('Do not call abstract method "showInfo()" from child.');
  }

}

class Jeep extends Car {

  constructor(fuel) {
    super();
    this.fuel = fuel;
  }

  // Implements abstract parent method
  showInfo() {
    console.log("\nModel: Jeep");
    console.log("Fuel: " + this.fuel.type());
  }

}

class Subaru extends Car {

  constructor(fuel) {
    super();
    this.fuel = fuel;
  }

  showInfo() {
    console.log("\nModel: Subaru");
    console.log("Fuel: " + this.fuel.type());
  }

}

console.log(":: Bridge ::\n");

const diesel = new Diesel();
const gasoline = new Gasoline();

const jeep = new Jeep(diesel);
const subaru = new Subaru(gasoline);

jeep.showInfo();
subaru.showInfo();